### PR TITLE
Bugfix/77 wpcom vip add role caps fix

### PR DESCRIPTION
**Pull request by ovidiul**
It fixes the issue mentioned in #77  by renaming the activation hook to admin_init and adding a version check to only init the roles once when the version changes as per https://docs.wpvip.com/how-tos/customize-user-roles/ .

PR #93 might need to be merged before this.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -4,3 +4,4 @@
 /phpcs.xml
 /phpunit.xml
 /.phpunit.result.cache
+node_modules
```
Comment (GaryJones) [inline .gitignore]: Unrelated change.

File: README.md [hunk 2]
```diff
@@ -21,6 +21,30 @@ Please contact us before using this plugin.
 - PHP 5.6+
 - WordPress 4.5+
 
+## Developement - Getting Started
```
Comment (GaryJones) [inline README.md]: Unrelated change.

File: composer.json [hunk 3]
```diff
@@ -1,5 +1,4 @@
-{
-  "name": "automattic/wpcom-legacy-redirector",
+{                                                                                                                                                                                                                              "name": "automattic/wpcom-legacy-redirector",
```
Comment (GaryJones) [inline composer.json]: Unrelated change.

File: composer.json [hunk 4]
```diff
@@ -41,3 +40,4 @@
     ]
   }
 }
+
```
Comment (GaryJones) [inline composer.json]: Don't include `composer.lock`.

File: includes/class-wpcom-legacy-redirector-cli.php [hunk 5]
```diff
@@ -351,13 +351,15 @@ public function export_to_csv( $args, $assoc_args ) {
...
 
 		do {
-			$posts = get_posts( array(
-				'posts_per_page'   => $posts_per_page,
-				'paged'            => $paged,
-				'post_type'        => WPCOM_Legacy_Redirector::POST_TYPE,
-				'post_status'      => 'any',
-				'suppress_filters' => 'false',
-			) );
+			$posts = get_posts(
```
Comment (GaryJones) [inline includes/class-wpcom-legacy-redirector-cli.php]: Unrelated change.

File: includes/class-wpcom-legacy-redirector-ui.php [hunk 6]
```diff
@@ -125,10 +125,25 @@ public function custom_vip_legacy_redirect_column( $column, $post_id ) {
 				break;
 		}
 	}
+
+	/**
+	 * Get WP Home URL without path suffix
+	 *
+	 * @return string
+	 */
+	private function get_home_domain_without_path() {
```
Comment (GaryJones) [inline includes/class-wpcom-legacy-redirector-ui.php]: Unrelated change - fork from `develop` and apply the changes to that new branch.

File: includes/class-wpcom-legacy-redirector.php [hunk 7]
```diff
@@ -6,13 +6,13 @@
 class WPCOM_Legacy_Redirector {
 	const POST_TYPE   = 'vip-legacy-redirect';
 	const CACHE_GROUP = 'vip-legacy-redirect-2';
-
+	const VERSION     = '1.4.0';
```
Comment (GaryJones) [inline includes/class-wpcom-legacy-redirector.php]: We've not tagged `1.4.0` yet, so this can't be correct.

File: includes/class-wpcom-legacy-redirector.php [hunk 8]
```diff
@@ -64,6 +64,14 @@ static function init() {
 	 * Register custom role using VIP Helpers with fallbacks.
 	 */
 	static function register_redirect_custom_capability() {
+
+		$plugin_version_key = self::POST_TYPE . '_version';
+
+		// Check if this has been run already.
+		if ( self::VERSION <= get_option( $plugin_version_key ) ) {
```
Comment (GaryJones) [inline includes/class-wpcom-legacy-redirector.php]: The use of the plugin version here means that it will redundantly run on each new plugin version. Instead, we can use a different constant that has an integer just related to the version of these roles and caps changes.

File: includes/class-wpcom-legacy-redirector.php [hunk 9]
```diff
@@ -328,7 +339,14 @@ static function get_redirect_uri( $url ) {
 				return add_query_arg( $preservable_params, get_permalink( $redirect_post->post_parent ) );
 			} elseif ( ! empty( $redirect_post->post_excerpt ) ) {
 				// Add preserved params to the destination URL.
-				return add_query_arg( $preservable_params, esc_url_raw( $redirect_post->post_excerpt ) );
+				// We need to add here the home_url() if the target starts with.
```
Comment (GaryJones) [inline includes/class-wpcom-legacy-redirector.php]: Unrelated change.

File: includes/class-wpcom-legacy-redirector.php [hunk 10]
```diff
@@ -365,6 +383,46 @@ private static function get_url_hash( $url ) {
...
+	 * @param string $url       The URL to parse.
+	 * @param int    $component The specific component to retrieve. Use one of the
+	 *                          PHP predefined constants to specify which one. Defaults
+	 *                          to -1 (= return all parts as an array).
+	 * @return array Exception on parse failure.
+	 *               Array of URL components on success; When a specific component has been
+	 *               requested: null if the component doesn't exist in the given URL; a
+	 *               string (or in the case of PHP_URL_PORT, integer) when it does. 
+	 */
+	public static function mb_parse_url( $url, $component = -1 ) {
```
Comment (GaryJones) [inline includes/class-wpcom-legacy-redirector.php]: Unrelated change.

File: package.json [hunk 11]
```diff
@@ -0,0 +1,26 @@
+{
```
Comment (GaryJones) [inline package.json]: Unrelated change.

File: tests/integration/RedirectsTest.php [hunk 12]
```diff
@@ -6,6 +6,23 @@ class RedirectsTest extends TestCase {
 
 	public function get_redirect_data() {
 		return array(
+			'redirect_relative_path'  => array(
+				'/non-existing-page',
```
Comment (GaryJones) [inline tests/integration/RedirectsTest.php]: Unrelated change.

File: wpcom-legacy-redirector.php [hunk 13]
```diff
@@ -25,7 +25,7 @@
 define( 'WPCOM_LEGACY_REDIRECTOR_PLUGIN_NAME', 'WPCOM Legacy Redirector' );
 
 if ( defined( 'WP_CLI' ) && WP_CLI ) {
-	require __DIR__ . '/includes/class-wpcom-legacy-redirector-cli.php';
+	include __DIR__ . '/includes/class-wpcom-legacy-redirector-cli.php';
```
Comment (GaryJones) [inline wpcom-legacy-redirector.php]: Unrelated change.

**Review by GaryJones — CHANGES_REQUESTED** [review]
There are too many unrelated changes in this PR. The actual fix needs about 4 lines of code, and a unit or integration test.

I have a local branch that extracts the logic for handling the capability into a new class, so I'll get that pushed and then we can add the logic and new constant there.

File: .gitignore (hunk 1 above)
Comment (ovidiul) [reply .gitignore]: removed

File: includes/class-wpcom-legacy-redirector.php (hunk 7 above)
Comment (ovidiul) [reply includes/class-wpcom-legacy-redirector.php]: fixed to latest known stable 1.3.0 for now

File: includes/class-wpcom-legacy-redirector.php (hunk 8 above)
Comment (ovidiul) [reply includes/class-wpcom-legacy-redirector.php]: I've added a new CAPABILITIES_VER constant for it to fix this with value set to 1.0

**Comment by ovidiul** [comment]
Hi Gary, I've rebased my PR and added the requested changes.

Let me know if you see further adjustments needed.

File: includes/class-wpcom-legacy-redirector.php [hunk 14]
```diff
@@ -13,27 +13,45 @@ class WPCOM_Legacy_Redirector {
 	const POST_TYPE   = 'vip-legacy-redirect';
 	const CACHE_GROUP = 'vip-legacy-redirect-2';
 
+	// Plugin version.
+	const PLUGIN_VERSION = '1.3.0';
+
+	// Used to flip the version of the available roles capabilities.
+	const CAPABILITIES_VER = '1.0';
```
Comment (GaryJones) [inline includes/class-wpcom-legacy-redirector.php]: You're doing a numerical comparison, but have this set as a string? Use an int instead.

```suggestion
	const CAPABILITIES_VER = 1;
```

File: includes/class-wpcom-legacy-redirector.php [hunk 15]
```diff
@@ -13,27 +13,45 @@ class WPCOM_Legacy_Redirector {
...
 		add_filter( 'bulk_actions-edit-' . Post_Type::POST_TYPE, array( __CLASS__, 'remove_bulk_edit' ) );
 	}
 
+	/**
+	 * Initialize and register other classes.
+	 */
+	static function admin_init() {
+		$capabilities_version_key = self::POST_TYPE . '_capabilities_version';
+
+		if ( self::CAPABILITIES_VER <= get_option( $capabilities_version_key ) ) {
```
Comment (GaryJones) [inline includes/class-wpcom-legacy-redirector.php]: I think this logic is wrong. Currently it says if the constant is less then or equal to the saved version number (which it should be for much of the time), then register the new capability.

I _think_ this should be:

```suggestion
		if ( self::CAPABILITIES_VER > get_option( $capabilities_version_key ) ) {
```

The integration tests below check that the capability is added to the roles, but it doesn't check how many times this `Capability->register()` command is run, which is part of what we're  trying to fix here.

File: tests/integration/CapabilityTest.php [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+
+	protected static $initialized = false;
+
+	public function setUp() {
+		if ( ! self::$initialized ) {
+			$capability = new Capability();
+			$capability->register();
+
+			self::$initialized = true;
+		}
```
Comment (GaryJones) [inline tests/integration/CapabilityTest.php]: To test the logic in `includes/class-wpcom-legacy-redirector.php` (about whether the Capability method is called or not), then you may want to have this moved to each method instead of in this setUp.

If anything is left in the setUp, then you may need a tearDown as well.

File: tests/integration/CapabilityTest.php [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+
+	/**
+	 * Test administrator capabilities for Capability::MANAGE_REDIRECTS_CAPABILITY
+	 *
+	 * @return void
+	 */
+	public function test_admin_capability() {
+
+		$user_id = $this->factory->user->create( array( 'role' => 'administrator' ) );
+		$user    = wp_set_current_user( $user_id );
```
Comment (GaryJones) [inline tests/integration/CapabilityTest.php]: Instead of testing a _new_ administrator, can you also check to see if an existing administrator account has got the cap? (Perhaps another reason to move the `Capability->register()` to these test methods so you can do a before and after check.

**Review by GaryJones — CHANGES_REQUESTED** [review]
I'd like to see more variety on the integration tests to confirm that the caps are not present on the various roles, then are added and are present, and that they are not re-added unnecessarily on subsequent requests as well.

File: includes/class-wpcom-legacy-redirector.php (hunk 15 above)
Comment (ovidiul) [reply includes/class-wpcom-legacy-redirector.php]: found a way to test this, I've moved related code to the capability class and added a return statement in that function, this would give us an easy way to test the register call after the initial init

File: includes/class-wpcom-legacy-redirector.php (hunk 14 above)
Comment (ovidiul) [reply includes/class-wpcom-legacy-redirector.php]: done

File: tests/integration/CapabilityTest.php (hunk 17 above)
Comment (ovidiul) [reply tests/integration/CapabilityTest.php]: I modified the tests accordingly, added also an unregister method with the relevant test in the Capability section.

File: tests/integration/CapabilityTest.php (hunk 16 above)
Comment (ovidiul) [reply tests/integration/CapabilityTest.php]: I've added the logic, but replaced the setUp with a beforeClass, we could skip this but I want to make sure we have no capabilities setup before the tests.

I use tearDown to check the register init and making sure it's not being setup again after the initial init.

File: tests/integration/CapabilityTest.php [hunk 18]
```diff
@@ -72,12 +86,37 @@ public function test_subscriber_capability() {
...
+
+		$this->assertTrue( array_key_exists( Capability::MANAGE_REDIRECTS_CAPABILITY, get_role( 'administrator' )->capabilities ) );
+
+		$capability->unregister();
+
+		$this->assertFalse( array_key_exists( Capability::MANAGE_REDIRECTS_CAPABILITY, get_role( 'administrator' )->capabilities ) );
+
+		$capability->register();
```
Comment (GaryJones) [inline tests/integration/CapabilityTest.php]: Is this final register really needed?

File: tests/integration/CapabilityTest.php [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+
+	/**
+	 * We remove the Capability::MANAGE_REDIRECTS_CAPABILITY  before testing
+	 *
+	 * @beforeClass
+	 */
+	public static function beforeClassSetup() {
+		$capability = new Capability();
+		$capability->unregister();
+	}
```
Comment (GaryJones) [inline tests/integration/CapabilityTest.php]: Is this needed? (It may be!)

As part of testing, we register the plugin as being activated, but that shouldn't trigger the `admin_init` hook which may do the cap registering.

If other test classes run first, I don't think it should be left in a caps-registered state.

If this method is scrapped, do the tests pass? If so, then it either it was redundant, or there's another thing to be tested.

File: tests/integration/CapabilityTest.php [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+
+	public function test_admin_capabilities_before_register() {
+		$user_id = $this->factory->user->create( array( 'role' => 'administrator' ) );
+		$user    = wp_set_current_user( $user_id );
```
Comment (GaryJones) [inline tests/integration/CapabilityTest.php]: As as separate test, can you grab user ID 1 and check their caps before and after registering?

So rename this test to `test_new_admin_capabilities_before_register` and create a `test_existiing_admin_capabilities_before_register` test method?

Comment (GaryJones) [reply tests/integration/CapabilityTest.php]: We'll also want to see tests for when `wpcom_vip_add_role_caps()` exists and when it doesn't since right now I suspect only one of those branches is being tested.

File: tests/integration/CapabilityTest.php [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+
+		$this->assertFalse( boolval( $user->has_cap( Capability::MANAGE_REDIRECTS_CAPABILITY ) ) );
```
Comment (GaryJones) [inline tests/integration/CapabilityTest.php]: `->has_cap()` only returns a boolean `true` or `false`, so `boolval()` is redundant here.

File: tests/integration/CapabilityTest.php [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+	 */
+	public function test_new_admin_capability() {
+
+		$capability        = new Capability();
+		$register_response = $capability->register();
+
+		$user_id = $this->factory->user->create( array( 'role' => 'administrator' ) );
+		$user    = wp_set_current_user( $user_id );
+
+		$this->assertTrue( $user->has_cap( Capability::MANAGE_REDIRECTS_CAPABILITY ) );
```
Comment (GaryJones) [inline tests/integration/CapabilityTest.php]: Let's wrap these assertions into custom assertion methods at the end of this test class. Something like:

```php
protected function assertUserHasRedirectCapability( $user ) {
	return $this->assertTrue( $user->has_cap( Capability::MANAGE_REDIRECTS_CAPABILITY ) );
}

protected function assertUserNotHasRedirectCapability( $user ) {
	return $this->assertFalse( $user->has_cap( Capability::MANAGE_REDIRECTS_CAPABILITY ) );
}

protected function assertRoleHasRedirectsCapability( $role ) {
	$user_id = $this->factory->user->create( array( 'role' => $role ) );
	$user    = wp_set_current_user( $user_id );

	return $this->assertUserHasRedirectCapability( $user );
}

protected function assertRoleNotHasRedirectsCapability( $role ) {
	$user_id = $this->factory->user->create( array( 'role' => $role ) );
	$user    = wp_set_current_user( $user_id );

	return $this->assertUserNotHasRedirectCapability( $user );
}
```

Then in the test methods, you can do:

```php
public function test_new_admin_capability() {
	$this->assertRoleNotHasRedirectsCapability( 'administrator' );

	$capability        = new Capability();
	$register_response = $capability->register();

	$this->assertRoleHasRedirectsCapability( 'administrator' );	
}
```

This could then be extracted to use a dataProvider since then the only variable is the role - wrap the final assertion in a condition that either checks the role has, or role has not got he cap, based on the expectation in the dataProvider.

Alternatively, just bundle the different role checks together:

```php
public function test_new_admin_capability() {
	$this->assertRoleNotHasRedirectsCapability( 'administrator' );
	$this->assertRoleNotHasRedirectsCapability( 'editor' );
	$this->assertRoleNotHasRedirectsCapability( 'subscriber' );

	$capability        = new Capability();
	$register_response = $capability->register();

	$this->assertRoleHasRedirectsCapability( 'administrator' );
	$this->assertRoleHasRedirectsCapability( 'editor' );
	$this->assertRoleNotHasRedirectsCapability( 'subscriber' ); // Should be no change.
}
```

**Review by GaryJones — CHANGES_REQUESTED** [review]

File: tests/integration/CapabilityTest.php (hunk 19 above)
Comment (ovidiul) [reply tests/integration/CapabilityTest.php]: It seems it is, what happens is that after each test, the wp_options tests table retains it's roles option to the mentioned state, so I think it's best after each test to remove any set capabilities

File: tests/integration/CapabilityTest.php (hunk 20 above)
Comment (ovidiul) [reply tests/integration/CapabilityTest.php]: I've added the ID=1 user test, but for multisite, this won't work because of in WP_User class, on multisite the has_cap [returns true](https://github.com/WordPress/WordPress/blob/master/wp-includes/class-wp-user.php#L754-L760) for admin users

Not sure about wpcom_vip_add_role_caps , we might need to mock it, but i think we should do that in a separate PR?

File: tests/integration/CapabilityTest.php (hunk 21 above)
Comment (ovidiul) [reply tests/integration/CapabilityTest.php]: This is removed now.

File: tests/integration/CapabilityTest.php (hunk 18 above)
Comment (ovidiul) [reply tests/integration/CapabilityTest.php]: Yes, it was needed for the tearDown, however I've refactored that now to not require it.

File: tests/integration/CapabilityTest.php (hunk 22 above)
Comment (ovidiul) [reply tests/integration/CapabilityTest.php]: New methods added and the necessary code refactored

File: tests/integration/CapabilityTest.php (hunk 18 above)
Comment (ovidiul) [reply tests/integration/CapabilityTest.php]: not anymore, it was needed for the tearDown method

File: includes/class-capability.php [hunk 23]
```diff
@@ -5,10 +5,21 @@
 final class Capability {
 	const MANAGE_REDIRECTS_CAPABILITY = 'manage_redirects';
 
+	// Used to flip the version of the available roles capabilities.
+	const CAPABILITIES_VER = '1';
```
Comment (GaryJones) [inline includes/class-capability.php]: This is still a string? It's used in line 19 with a `<=` operator.

File: includes/class-capability.php [hunk 24]
```diff
@@ -19,5 +30,34 @@ public function register() {
...
+	}
+
+	/**
+	 * Unregister the capabilities
+	 *
+	 * @return void
+	 */
+	public function unregister() {
+
+		$capabilities_version_key = self::MANAGE_REDIRECTS_CAPABILITY . '_capability_version';
```
Comment (GaryJones) [inline includes/class-capability.php]: This is the second time this line is used. Let's extract it into its own method for consistency.

File: includes/class-wpcom-legacy-redirector.php [hunk 25]
```diff
@@ -13,27 +13,36 @@ class WPCOM_Legacy_Redirector {
...
 		add_filter( 'admin_enqueue_scripts', array( __CLASS__, 'wpcom_legacy_add_redirect_js' ) );
 		add_filter( 'bulk_actions-edit-' . Post_Type::POST_TYPE, array( __CLASS__, 'remove_bulk_edit' ) );
 	}
 
+	/**
+	 * Initialize and register other classes.
+	 */
+	static function admin_init() {
+		$capability = new Capability();
+			$capability->register();
```
Comment (GaryJones) [inline includes/class-wpcom-legacy-redirector.php]: Incorrect number of indentation tabs.

File: tests/integration/CapabilityTest.php [hunk 26]
```diff
@@ -0,0 +1,130 @@
...
+	/**
+	 * Test if the capabilities register function fires again after init
+	 *
+	 * @return void
+	 */
+	public function tearDown() {
+
+		$capability = new Capability();
+		$capability->unregister();
+	}
```
Comment (GaryJones) [inline tests/integration/CapabilityTest.php]: Move this fixture to the top, and check the DocBlock description.

**Review by GaryJones — CHANGES_REQUESTED** [review]
Looking better, but I'm still not seeing a test that checks that update_option() was only called once when `$capability->register()` called twice. Have a look at https://brain-wp.github.io/BrainMonkey/docs/functions-expect.html which allows you to test for exactly this. Set the `Function\expects`, ensure the saved option is empty, call register twice, and see if the test is green (Might need to put that into a unit test, rather than the integration test)

File: includes/class-capability.php (hunk 23 above)
Comment (ovidiul) [reply includes/class-capability.php]: Fixed now

File: includes/class-capability.php (hunk 24 above)
Comment (ovidiul) [reply includes/class-capability.php]: done

File: includes/class-wpcom-legacy-redirector.php (hunk 25 above)
Comment (ovidiul) [reply includes/class-wpcom-legacy-redirector.php]: fixed

File: tests/integration/CapabilityTest.php (hunk 26 above)
Comment (ovidiul) [reply tests/integration/CapabilityTest.php]: fixed

**Comment by ovidiul** [comment]
> Looking better, but I'm still not seeing a test that checks that update_option() was only called once when `$capability->register()` called twice. Have a look at [brain-wp.github.io/BrainMonkey/docs/functions-expect.html](https://brain-wp.github.io/BrainMonkey/docs/functions-expect.html) which allows you to test for exactly this. Set the `Function\expects`, ensure the saved option is empty, call register twice, and see if the test is green (Might need to put that into a unit test, rather than the integration test)

I've added a unit test in tests/unit/CapabilityTest.php

We do have a related test in `tests/integration/CapabilityTest.php` , method `test_register` which checks the return of the register method, I guess another one doesn't hurt.

File: tests/unit/CapabilityTest.php [hunk 27]
```diff
@@ -0,0 +1,65 @@
...
+
+	/**
+	 * TEar down after each test
+	 *
+	 * @return void
+	 */
+	protected function tearDown() {
+		Monkey\tearDown();
+		parent::tearDown();
+	}
```
Comment (GaryJones) [inline tests/unit/CapabilityTest.php]: Because this class extends `Yoast\WPTestUtils\BrainMonkey\TestCase` then I believe these Monkey setup and teardown is already handled - see https://github.com/Yoast/wp-test-utils/blob/develop/src/BrainMonkey/TestCase.php

**Review by GaryJones — CHANGES_REQUESTED** [review]

File: tests/unit/CapabilityTest.php (hunk 27 above)
Comment (ovidiul) [reply tests/unit/CapabilityTest.php]: even better, i've removed that now.

File: composer.json [hunk 28]
```diff
@@ -16,7 +16,8 @@
   },
   "require-dev": {
     "phpunit/phpunit": "^5 || ^6 || ^7",
-    "yoast/wp-test-utils": "^0.2"
+    "yoast/wp-test-utils": "^0.2",
+    "brain/monkey": "^2.6"
```
Comment (GaryJones) [inline composer.json]: Brain Monkey is already included with wp-test-utils.

**Comment by GaryJones** [comment]
I've squashed all of the commits, and rebased on to `develop`, so we'll have a clean history.

**Review by GaryJones — APPROVED** [review]
LGTM!
